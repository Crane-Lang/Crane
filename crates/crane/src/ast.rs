use smol_str::SmolStr;
use thin_vec::ThinVec;

/// An identifier.
#[derive(Debug, Clone)]
pub struct Ident(pub SmolStr);

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

/// The kind of a Crane expression.
#[derive(Debug, Clone)]
pub enum ExprKind {
    /// A reference to a variable.
    Variable { name: Ident },

    /// A function call.
    Call {
        fun: Box<Expr>,
        args: ThinVec<Box<Expr>>,
    },
}

/// A Crane expression.
#[derive(Debug, Clone)]
pub struct Expr {
    pub kind: ExprKind,
}

/// The kind of a Crane statement.
#[derive(Debug, Clone)]
pub enum StmtKind {
    /// A function declaration (`fn`).
    Fn(Box<Fn>),

    /// An expression.
    Expr(Expr),
}

/// A function definition.
#[derive(Debug, Clone)]
pub struct Fn {
    pub name: SmolStr,
    pub body: ThinVec<Stmt>,
}

/// A Crane statement.
#[derive(Debug, Clone)]
pub struct Stmt {
    pub kind: StmtKind,
}
