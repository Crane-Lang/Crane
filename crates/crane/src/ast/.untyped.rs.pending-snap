{"run_id":"1687112702-689890000","line":225,"new":{"module_name":"crane__ast__untyped__tests","snapshot_name":"ast_node_sizes-4","metadata":{"source":"crates/crane/src/ast/untyped.rs","assertion_line":225,"expression":"size_of::<Item>().to_string()"},"snapshot":"64"},"old":{"module_name":"crane__ast__untyped__tests","metadata":{},"snapshot":"56"}}
{"run_id":"1687112777-693774000","line":222,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":223,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":224,"new":null,"old":null}
{"run_id":"1687112777-693774000","line":225,"new":{"module_name":"crane__ast__untyped__tests","snapshot_name":"ast_node_sizes-4","metadata":{"source":"crates/crane/src/ast/untyped.rs","assertion_line":225,"expression":"size_of::<Item>().to_string()"},"snapshot":"64"},"old":{"module_name":"crane__ast__untyped__tests","metadata":{},"snapshot":"56"}}
{"run_id":"1687112794-630697000","line":222,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":223,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":224,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":225,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":226,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":227,"new":null,"old":null}
{"run_id":"1687112794-630697000","line":228,"new":null,"old":null}
