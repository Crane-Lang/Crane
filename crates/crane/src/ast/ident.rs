use smol_str::SmolStr;

// TODO: Add span to `Ident`.
/// An identifier.
#[derive(Debug, PartialEq, Eq, Hash, Clone)]
pub struct Ident(pub SmolStr);

impl std::fmt::Display for Ident {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
