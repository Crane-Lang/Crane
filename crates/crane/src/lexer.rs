mod error;

pub mod token;

pub use error::*;

use logos::Logos;

use crate::lexer::token::{Token, TokenKind};

pub struct Lexer<'src> {
    lexer: logos::Lexer<'src, TokenKind>,
}

impl<'src> Lexer<'src> {
    pub fn new(input: &'src str) -> Self {
        Self {
            lexer: TokenKind::lexer(input),
        }
    }
}

impl<'src> Iterator for Lexer<'src> {
    type Item = Result<Token, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        let kind = self.lexer.next()?;
        let lexeme = self.lexer.slice();
        let span = self.lexer.span().into();

        Some(
            kind.map(|kind| Token {
                kind,
                lexeme: lexeme.into(),
                span,
            })
            .map_err(|kind| LexError { kind, span }),
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn lex(input: &str) -> logos::Lexer<TokenKind> {
        TokenKind::lexer(input)
    }

    fn check(input: &str, token: TokenKind) {
        let mut lexer = TokenKind::lexer(input);

        assert_eq!(lexer.next(), Some(Ok(token)));
        assert_eq!(lexer.slice(), input);
    }

    #[test]
    fn lex_open_paren() {
        check("(", TokenKind::OpenParen)
    }

    #[test]
    fn lex_close_paren() {
        check(")", TokenKind::CloseParen)
    }

    #[test]
    fn lex_open_brace() {
        check("{", TokenKind::OpenBrace)
    }

    #[test]
    fn lex_close_brace() {
        check("}", TokenKind::CloseBrace)
    }

    #[test]
    fn lex_comment() {
        check("// This is a comment.", TokenKind::Comment)
    }

    #[test]
    fn lex_fn_declaration() {
        let mut lex = lex("fn foo() {}");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 0..2);
        assert_eq!(lex.slice(), "fn");

        assert_eq!(lex.next(), Some(Ok(TokenKind::Ident)));
        assert_eq!(lex.span(), 3..6);
        assert_eq!(lex.slice(), "foo");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenParen)));
        assert_eq!(lex.span(), 6..7);
        assert_eq!(lex.slice(), "(");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseParen)));
        assert_eq!(lex.span(), 7..8);
        assert_eq!(lex.slice(), ")");

        assert_eq!(lex.next(), Some(Ok(TokenKind::OpenBrace)));
        assert_eq!(lex.span(), 9..10);
        assert_eq!(lex.slice(), "{");

        assert_eq!(lex.next(), Some(Ok(TokenKind::CloseBrace)));
        assert_eq!(lex.span(), 10..11);
        assert_eq!(lex.slice(), "}");
    }

    #[test]
    fn test_lexer() {
        insta::glob!("snapshot_inputs/*.crane", |path| {
            let source = std::fs::read_to_string(path).unwrap();

            let lexer = Lexer::new(&source);

            insta::assert_yaml_snapshot!(lexer.into_iter().collect::<Vec<_>>());
        })
    }
}
