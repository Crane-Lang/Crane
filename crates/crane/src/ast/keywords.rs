use smol_str::SmolStr;

use crate::ast::{Ident, DUMMY_SPAN};

pub const FN: Ident = Ident {
    name: SmolStr::new_inline("fn"),
    span: DUMMY_SPAN,
};

pub const LET: Ident = Ident {
    name: SmolStr::new_inline("let"),
    span: DUMMY_SPAN,
};

pub const MOD: Ident = Ident {
    name: SmolStr::new_inline("mod"),
    span: DUMMY_SPAN,
};

pub const PUB: Ident = Ident {
    name: SmolStr::new_inline("pub"),
    span: DUMMY_SPAN,
};

pub const STRUCT: Ident = Ident {
    name: SmolStr::new_inline("struct"),
    span: DUMMY_SPAN,
};

pub const UNION: Ident = Ident {
    name: SmolStr::new_inline("union"),
    span: DUMMY_SPAN,
};
