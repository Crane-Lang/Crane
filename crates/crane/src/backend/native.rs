use std::process::Command;

use inkwell::builder::Builder;
use inkwell::context::Context;
use inkwell::module::{Linkage, Module};
use inkwell::passes::PassManager;
use inkwell::targets::{
    CodeModel, FileType, InitializationConfig, RelocMode, Target, TargetTriple,
};
use inkwell::types::BasicType;
use inkwell::values::{
    BasicMetadataValueEnum, BasicValue, CallSiteValue, FunctionValue, GlobalValue, IntValue,
};
use inkwell::{AddressSpace, OptimizationLevel};
use thin_vec::ThinVec;

use crate::ast::{Expr, ExprKind, FnParam, Item, ItemKind, Literal, LiteralKind, StmtKind};

pub struct NativeBackend {
    context: Context,
}

impl NativeBackend {
    pub fn new() -> Self {
        Self {
            context: Context::create(),
        }
    }

    pub fn compile(&self, program: Vec<Item>) {
        Target::initialize_aarch64(&InitializationConfig::default());

        let opt = OptimizationLevel::Default;
        let reloc = RelocMode::Default;
        let model = CodeModel::Default;

        let target = Target::from_name("aarch64").expect("Failed to parse target");

        let target = dbg!(target);

        let target_machine = target
            .create_target_machine(
                &TargetTriple::create("aarch64-apple-darwin"),
                "apple-m2",
                "",
                opt,
                reloc,
                model,
            )
            .unwrap();

        let target_machine = dbg!(target_machine);

        let module = self.context.create_module("main");
        let builder = self.context.create_builder();

        let fpm = PassManager::create(&module);

        fpm.add_instruction_combining_pass();

        fpm.initialize();

        // Define `puts`.
        {
            let i8_type = self.context.i8_type();
            let i32_type = self.context.i32_type();
            let fn_type = i32_type.fn_type(
                &[i8_type
                    .ptr_type(AddressSpace::default())
                    .as_basic_type_enum()
                    .into()],
                false,
            );

            let puts = module.add_function("puts", fn_type, Some(Linkage::External));

            dbg!(puts);
        }

        // Define `sprintf`.
        {
            let i8_type = self.context.i8_type();
            let i32_type = self.context.i32_type();

            let fn_type = i32_type.fn_type(
                &[
                    i8_type
                        .ptr_type(AddressSpace::default())
                        .as_basic_type_enum()
                        .into(),
                    i8_type
                        .ptr_type(AddressSpace::default())
                        .as_basic_type_enum()
                        .into(),
                ],
                true,
            );

            let sprintf = module.add_function("sprintf", fn_type, Some(Linkage::External));

            dbg!(sprintf);
        }

        // Define `printf`.
        {
            let i8_type = self.context.i8_type();
            let i32_type = self.context.i32_type();

            let fn_type = i32_type.fn_type(
                &[i8_type
                    .ptr_type(AddressSpace::default())
                    .as_basic_type_enum()
                    .into()],
                true,
            );

            let printf = module.add_function("printf", fn_type, Some(Linkage::External));

            dbg!(printf);
        }

        // Define `print`.
        {
            let fn_name = "print";

            let i8_type = self.context.i8_type();

            let fn_type = self.context.void_type().fn_type(
                &[i8_type
                    .ptr_type(AddressSpace::default())
                    .as_basic_type_enum()
                    .into()],
                false,
            );

            let fn_value = module.add_function(&fn_name, fn_type, None);

            let value_param = fn_value.get_first_param().unwrap();

            let entry = self.context.append_basic_block(fn_value, "entry");

            builder.position_at_end(entry);

            let template = b"%1$s";

            let i8_type = self.context.i8_type();
            let i8_array_type = i8_type.array_type(template.len() as u32 + 1);

            let template = self.context.const_string(template, true);

            let global = module.add_global(i8_array_type, None, "print_template");
            global.set_linkage(Linkage::Internal);
            global.set_constant(true);
            global.set_initializer(&template);

            if let Some(callee) = module.get_function(&"printf") {
                builder.build_call(
                    callee,
                    &[global.as_basic_value_enum().into(), value_param.into()],
                    "tmp",
                );
            } else {
                eprintln!("Function '{}' not found.", "puts");
            }

            builder.build_return(None);

            if fn_value.verify(true) {
                fpm.run_on(&fn_value);

                println!("{} is verified!", fn_name);
            } else {
                println!("{} is not verified :(", fn_name);
            }

            dbg!(fn_value);
        }

        // Define `println`.
        {
            let fn_name = "println";

            let i8_type = self.context.i8_type();

            let fn_type = self.context.void_type().fn_type(
                &[i8_type
                    .ptr_type(AddressSpace::default())
                    .as_basic_type_enum()
                    .into()],
                false,
            );

            let fn_value = module.add_function(&fn_name, fn_type, None);

            let value_param = fn_value.get_first_param().unwrap();

            let entry = self.context.append_basic_block(fn_value, "entry");

            builder.position_at_end(entry);

            if let Some(callee) = module.get_function(&"puts") {
                builder.build_call(callee, &[value_param.into()], "tmp");
            } else {
                eprintln!("Function '{}' not found.", "puts");
            }

            builder.build_return(None);

            if fn_value.verify(true) {
                fpm.run_on(&fn_value);

                println!("{} is verified!", fn_name);
            } else {
                println!("{} is not verified :(", fn_name);
            }

            dbg!(fn_value);
        }

        // Define `int_add`.
        {
            let fn_name = "int_add";

            let i128_type = self.context.i128_type();

            let fn_type = self.context.i128_type().fn_type(
                &[
                    i128_type.as_basic_type_enum().into(),
                    i128_type.as_basic_type_enum().into(),
                ],
                false,
            );

            let fn_value = module.add_function(&fn_name, fn_type, None);

            let lhs_param = fn_value.get_first_param().unwrap().into_int_value();
            let rhs_param = fn_value.get_nth_param(1).unwrap().into_int_value();

            let entry = self.context.append_basic_block(fn_value, "entry");

            builder.position_at_end(entry);

            let sum = builder.build_int_add(lhs_param, rhs_param, "sum");

            builder.build_return(Some(&sum));

            if fn_value.verify(true) {
                fpm.run_on(&fn_value);

                println!("{} is verified!", fn_name);
            } else {
                println!("{} is not verified :(", fn_name);
            }

            dbg!(fn_value);
        }

        // Define `int_to_string`.
        {
            let fn_name = "int_to_string";

            let i128_type = self.context.i128_type();
            let i8_type = self.context.i8_type();
            let i8_ptr_type = i8_type.ptr_type(AddressSpace::default());

            let fn_type = i8_ptr_type.fn_type(&[i128_type.as_basic_type_enum().into()], false);

            let fn_value = module.add_function(&fn_name, fn_type, None);

            let int_value = fn_value.get_first_param().unwrap().into_int_value();

            let entry = self.context.append_basic_block(fn_value, "entry");

            builder.position_at_end(entry);

            let buffer = builder
                .build_malloc(i8_ptr_type, "buffer")
                .expect("Failed to allocate `int_to_string` buffer.");

            let template = b"%1$d";

            let i8_type = self.context.i8_type();
            let i8_array_type = i8_type.array_type(template.len() as u32 + 1);

            let template = self.context.const_string(template, true);

            let global = module.add_global(i8_array_type, None, "int_to_string_template");
            global.set_linkage(Linkage::Internal);
            global.set_constant(true);
            global.set_initializer(&template);

            if let Some(callee) = module.get_function(&"sprintf") {
                builder.build_call(
                    callee,
                    &[
                        buffer.into(),
                        global.as_basic_value_enum().into(),
                        int_value.into(),
                    ],
                    "tmp",
                );
            } else {
                panic!("Function '{}' not found.", "sprintf");
            }

            builder.build_return(Some(&buffer));

            if fn_value.verify(true) {
                fpm.run_on(&fn_value);

                println!("{} is verified!", fn_name);
            } else {
                println!("{} is not verified :(", fn_name);
            }

            dbg!(fn_value);
        }

        for item in program
            // HACK: Reverse the items so we define the helper functions before `main`.
            // This should be replaced with a call graph.
            .into_iter()
            .rev()
        {
            match item.kind {
                ItemKind::Fn(fun) => {
                    let params = fun
                        .params
                        .iter()
                        .map(|_param| {
                            let i8_type = self.context.i8_type();

                            i8_type
                                .ptr_type(AddressSpace::default())
                                .as_basic_type_enum()
                                .into()
                        })
                        .collect::<Vec<_>>();

                    let fn_type = self.context.void_type().fn_type(&params, false);

                    let fn_value = module.add_function(&item.name.to_string(), fn_type, None);

                    for (index, param_value) in fn_value.get_param_iter().enumerate() {
                        if let Some(param) = fun.params.get(index) {
                            param_value.set_name(&param.name.to_string());
                        }
                    }

                    let entry = self.context.append_basic_block(fn_value, "entry");

                    builder.position_at_end(entry);

                    for stmt in fun.body {
                        match stmt.kind {
                            StmtKind::Expr(expr) => Self::compile_expr(
                                &self.context,
                                &builder,
                                &module,
                                &fun.params,
                                &fn_value,
                                expr,
                            ),
                            StmtKind::Item(item) => todo!(),
                        }
                    }

                    builder.build_return(None);

                    dbg!(fn_value);

                    if fn_value.verify(true) {
                        fpm.run_on(&fn_value);

                        println!("{} is verified!", item.name);
                    } else {
                        println!("{} is not verified :(", item.name);
                    }

                    fn_value.print_to_stderr();
                }
            }
        }

        dbg!(module.get_functions().count());

        module
            .print_to_file("build/main.ll")
            .expect("Failed to emit main.ll");

        let buffer = target_machine
            .write_to_memory_buffer(&module, FileType::Object)
            .expect("Failed to write to buffer");

        use std::io::Write;

        let mut outfile = std::fs::File::create("build/main.o").unwrap();

        outfile.write_all(buffer.as_slice()).unwrap();

        let bitcode = module.write_bitcode_to_memory();

        outfile.write_all(bitcode.as_slice()).unwrap();

        let exit_status = Command::new("clang")
            .args(["-o", "build/main", "build/main.o"])
            .status()
            .expect("Failed to build with clang");

        println!("clang exited with {}", exit_status);
    }

    fn compile_expr<'ctx>(
        context: &'ctx Context,
        builder: &Builder<'ctx>,
        module: &Module<'ctx>,
        fn_params: &ThinVec<FnParam>,
        fn_value: &FunctionValue<'ctx>,
        expr: Expr,
    ) {
        match expr.kind {
            ExprKind::Literal(literal) => match literal.kind {
                LiteralKind::String => {
                    Self::compile_string_literal(&context, &builder, &module, literal);
                }
                LiteralKind::Integer => {}
            },
            ExprKind::Variable { name } => todo!(),
            ExprKind::Call { fun, args } => {
                Self::compile_fn_call(
                    context,
                    builder,
                    module,
                    fn_value,
                    fn_params,
                    fun.clone(),
                    args,
                )
                .expect(&format!("Failed to compile function call: {:?}", fun));
            }
        }
    }

    fn compile_string_literal<'ctx>(
        context: &'ctx Context,
        builder: &Builder<'ctx>,
        module: &Module<'ctx>,
        literal: Literal,
    ) -> GlobalValue<'ctx> {
        // Unquote the string literal.
        let value = {
            let mut chars = literal.value.chars();
            chars.next();
            chars.next_back();
            chars.as_str()
        };

        let value = value.as_bytes();

        let i8_type = context.i8_type();
        let i8_array_type = i8_type.array_type(value.len() as u32 + 1);

        let string = context.const_string(value, true);

        let global = module.add_global(i8_array_type, None, "string_lit");
        global.set_linkage(Linkage::Internal);
        global.set_constant(true);
        global.set_initializer(&string);

        global
    }

    fn compile_integer_literal<'ctx>(
        context: &'ctx Context,
        builder: &Builder<'ctx>,
        module: &Module<'ctx>,
        literal: Literal,
    ) -> IntValue<'ctx> {
        let int_type = match literal.kind {
            LiteralKind::Integer => context.i128_type(),
            LiteralKind::String => unreachable!("Tried to compile a string literal as an integer."),
        };

        int_type.const_int(literal.value.parse().unwrap(), false)
    }

    fn compile_fn_call<'ctx>(
        context: &'ctx Context,
        builder: &Builder<'ctx>,
        module: &Module<'ctx>,
        caller: &FunctionValue<'ctx>,
        caller_params: &ThinVec<FnParam>,
        fun: Box<Expr>,
        args: ThinVec<Box<Expr>>,
    ) -> Result<CallSiteValue<'ctx>, String> {
        let callee_name = match fun.kind {
            ExprKind::Variable { name } => name,
            _ => todo!(),
        };

        if let Some(callee) = module.get_function(&callee_name.to_string()) {
            let args: Vec<BasicMetadataValueEnum> = args
                .into_iter()
                .map(|arg| match arg.kind {
                    ExprKind::Literal(literal) => match literal.kind {
                        LiteralKind::String => {
                            Self::compile_string_literal(&context, &builder, &module, literal)
                                .as_basic_value_enum()
                                .into()
                        }
                        LiteralKind::Integer => {
                            Self::compile_integer_literal(&context, &builder, &module, literal)
                                .as_basic_value_enum()
                                .into()
                        }
                    },
                    ExprKind::Variable { name } => {
                        let (param_index, _) = caller_params
                            .into_iter()
                            .enumerate()
                            .find(|(_, param)| param.name == name)
                            .expect(&format!("Param '{}' not found.", name));

                        caller
                            .get_nth_param(param_index as u32)
                            .expect("Param not found")
                            .as_basic_value_enum()
                            .into()
                    }
                    ExprKind::Call { fun, args } => Self::compile_fn_call(
                        &context,
                        &builder,
                        &module,
                        caller,
                        caller_params,
                        fun,
                        args,
                    )
                    .unwrap()
                    .try_as_basic_value()
                    .unwrap_left()
                    .into(),
                })
                .collect::<Vec<_>>();

            Ok(builder.build_call(callee, args.as_slice(), "tmp"))
        } else {
            eprintln!("Function '{}' not found.", callee_name);
            Err(format!("Function '{}' not found.", callee_name))
        }
    }
}
