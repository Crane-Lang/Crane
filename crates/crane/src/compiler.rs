use std::io::Write;
use std::path::PathBuf;

use ariadne::{Color, Label, Report, ReportKind, Source};
use itertools::Itertools;
use thin_vec::thin_vec;

use crate::ast::{Module, Package, SourceSpan};
use crate::backend::native::NativeBackend;
use crate::lexer::Lexer;
use crate::parser::{ParseErrorKind, Parser};
use crate::typer::{TypeErrorKind, Typer};

/// The input to the compiler.
pub enum Input {
    /// Load the source code from a file.
    File(PathBuf),

    /// Load source code from a string.
    String {
        /// The "filename" where the source code is coming from.
        filename: String,

        /// The source code.
        input: String,
    },
}

pub struct CompileParams {
    pub input: Input,
}

/// The interface to the Crane compiler.
pub struct Compiler {}

impl Compiler {
    pub fn new() -> Self {
        Self {}
    }

    pub fn compile<'io>(
        &mut self,
        stderr: &'io mut impl Write,
        params: CompileParams,
    ) -> Result<(), ()> {
        let (filepath, source) = match params.input {
            Input::File(path) => (
                path.display().to_string(),
                std::fs::read_to_string(&path).unwrap(),
            ),
            Input::String { filename, input } => (filename, input),
        };

        let lexer = Lexer::new(&source);
        let parser = Parser::new(lexer);

        match parser.parse() {
            Ok(items) => {
                let mut typer = Typer::new();

                let module = Module { items };

                let package = Package {
                    modules: thin_vec![module],
                };

                match typer.type_check_package(package) {
                    Ok(typed_package) => {
                        std::fs::create_dir_all("build").unwrap();

                        let context = inkwell::context::Context::create();

                        let backend = NativeBackend::new(&context);

                        backend.compile(typed_package);

                        println!("Compiled!");

                        Ok(())
                    }
                    Err(type_error) => {
                        let span = type_error.span;

                        let error_report = match type_error.kind {
                            TypeErrorKind::InvalidFunctionName { reason, suggestion } => {
                                Report::build(ReportKind::Error, &filepath, 1)
                                    .with_message("A type error occurred.")
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(reason)
                                            .with_color(Color::Red),
                                    )
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(format!(
                                                "Try writing it as `{suggestion}` instead."
                                            ))
                                            .with_color(Color::Cyan),
                                    )
                                    .finish()
                            }
                            TypeErrorKind::InvalidTypeName { reason, suggestion } => {
                                Report::build(ReportKind::Error, &filepath, 1)
                                    .with_message("A type error occurred.")
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(reason)
                                            .with_color(Color::Red),
                                    )
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(format!(
                                                "Try writing it as `{suggestion}` instead."
                                            ))
                                            .with_color(Color::Cyan),
                                    )
                                    .finish()
                            }
                            TypeErrorKind::UnknownModule { path, options } => {
                                let report = Report::build(ReportKind::Error, &filepath, 1)
                                    .with_message("A type error occurred.")
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(format!(
                                                "Module `{path}` does not exist.",
                                            ))
                                            .with_color(Color::Red),
                                    );

                                let suggestion = options
                                    .iter()
                                    .sorted_by_key(|option| option.to_string())
                                    .min_by_key(|option| {
                                        strsim::levenshtein(&option.to_string(), &path.to_string())
                                    });

                                let report = if let Some(suggestion) = suggestion {
                                    report.with_label(
                                        Label::new(SourceSpan::from((&filepath, suggestion.span)))
                                            .with_message(format!(
                                                "There is a module with a similar name: `{}`.",
                                                suggestion.clone()
                                            ))
                                            .with_color(Color::Cyan),
                                    )
                                } else {
                                    report
                                };

                                report.finish()
                            }
                            TypeErrorKind::UnknownFunction { path, options } => {
                                let report = Report::build(ReportKind::Error, &filepath, 1)
                                    .with_message("A type error occurred.")
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(format!(
                                                "Function `{path}` does not exist.",
                                            ))
                                            .with_color(Color::Red),
                                    );

                                let suggestion = options
                                    .iter()
                                    .sorted_by_key(|option| option.to_string())
                                    .min_by_key(|option| {
                                        strsim::levenshtein(&option.to_string(), &path.to_string())
                                    });

                                let report = if let Some(suggestion) = suggestion {
                                    report.with_label(
                                        Label::new(SourceSpan::from((&filepath, suggestion.span)))
                                            .with_message(format!(
                                                "There is a function with a similar name: `{}`.",
                                                suggestion.clone()
                                            ))
                                            .with_color(Color::Cyan),
                                    )
                                } else {
                                    report
                                };

                                report.finish()
                            }
                            TypeErrorKind::UnknownType { path, options } => {
                                let report = Report::build(ReportKind::Error, &filepath, 1)
                                    .with_message("A type error occurred.")
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(format!("Type `{path}` does not exist.",))
                                            .with_color(Color::Red),
                                    );

                                let suggestion = options
                                    .iter()
                                    .sorted_by_key(|option| option.to_string())
                                    .min_by_key(|option| {
                                        strsim::levenshtein(&option.to_string(), &path.to_string())
                                    });

                                let report = if let Some(suggestion) = suggestion {
                                    report.with_label(
                                        Label::new(SourceSpan::from((&filepath, suggestion.span)))
                                            .with_message(format!(
                                                "There is a type with a similar name: `{}`.",
                                                suggestion.clone()
                                            ))
                                            .with_color(Color::Cyan),
                                    )
                                } else {
                                    report
                                };

                                report.finish()
                            }
                            TypeErrorKind::Error(message) => {
                                Report::build(ReportKind::Error, &filepath, 1)
                                    .with_message("A type error occurred.")
                                    .with_label(
                                        Label::new(SourceSpan::from((&filepath, span)))
                                            .with_message(message)
                                            .with_color(Color::Red),
                                    )
                                    .finish()
                            }
                        };

                        error_report
                            .write((filepath.into(), Source::from(source)), stderr)
                            .unwrap();

                        Err(())
                    }
                }
            }
            Err(err) => {
                let span = err.span;

                let error_report = match err.kind {
                    ParseErrorKind::LexError(lex_error) => {
                        Report::build(ReportKind::Error, &filepath, 1)
                            .with_message("An error occurred during lexing.")
                            .with_label(
                                Label::new(SourceSpan::from((&filepath, span)))
                                    .with_message(lex_error)
                                    .with_color(Color::Red),
                            )
                            .finish()
                    }
                    ParseErrorKind::Error(message) => {
                        Report::build(ReportKind::Error, &filepath, 1)
                            .with_message("An error occurred during parsing.")
                            .with_label(
                                Label::new(SourceSpan::from((&filepath, span)))
                                    .with_message(message)
                                    .with_color(Color::Red),
                            )
                            .finish()
                    }
                };

                error_report
                    .write((filepath.into(), Source::from(source)), stderr)
                    .unwrap();

                Err(())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn test_camel_case_function_name() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "camel_case.crane".into(),
                input: r#"
fn main() {}

fn inCamelCase() {}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }

    #[test]
    pub fn test_mixed_case_function_name() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "mixed_case.crane".into(),
                input: r#"
fn main() {}

fn XMLHttpRequest() {}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }

    #[test]
    pub fn test_snake_case_struct_name() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "snake_case.crane".into(),
                input: r#"
struct snake_cased_struct {
    foo: Uint64,
    bar: Uint64,
}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }

    #[test]
    pub fn test_mixed_case_struct_name() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "mixed_case.crane".into(),
                input: r#"
struct XMLHttpRequest {
    foo: Uint64,
    bar: Uint64,
}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }

    #[test]
    pub fn test_snake_case_union_name() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "snake_case.crane".into(),
                input: r#"
union snake_cased_union {
    Foo,
    Bar,
}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }

    #[test]
    pub fn test_mixed_case_union_name() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "mixed_case.crane".into(),
                input: r#"
union XMLHttpRequest {
    Foo,
    Bar,
}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }

    #[test]
    pub fn test_struct_expression() {
        let mut compiler = Compiler::new();

        let params = CompileParams {
            input: Input::String {
                filename: "mixed_case.crane".into(),
                input: r#"
struct User {
    first_name: String,
    last_name: String,
    age: Uint64,
}

fn main() {
    let user = User {
        first_name: "Elaine",
        last_name: "Benes",
        age: 27,
    }
}
                "#
                .trim()
                .to_string(),
            },
        };

        let mut stderr = Vec::new();

        let _ = compiler.compile(&mut stderr, params);

        let stderr = strip_ansi_escapes::strip(stderr).unwrap();
        let stderr = std::str::from_utf8(&stderr).unwrap();

        insta::assert_snapshot!(&stderr);
    }
}
