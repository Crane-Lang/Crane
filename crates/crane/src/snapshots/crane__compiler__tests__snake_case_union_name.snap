---
source: crates/crane/src/compiler.rs
expression: "&stderr"
---
Error: A type error occurred.
   ╭─[snake_case.crane:1:2]
   │
 1 │ union snake_cased_union {
   │       ────────┬────────  
   │               ╰────────── Union names must be written in PascalCase.
   │               │          
   │               ╰────────── Try writing it as `SnakeCasedUnion` instead.
───╯

